lib/Doctrine/DBAL/TransactionManager.php

```diff
@@ -0,0 +1,284 @@
+<?php
+
+namespace Doctrine\DBAL;
+
+/**
+ * Manages transactions.
+ */
+class TransactionManager
+{
+    const SAVEPOINT_PREFIX = 'DOCTRINE2_SAVEPOINT_';
+
+    /**
+     * @var \Doctrine\DBAL\Connection
+     */
+    private $connection;
+
+    /**
+     * @var \Doctrine\DBAL\Transaction[]
+     */
+    private $activeTransactions = array();
+
+    /**
+     * Whether nested transactions should use savepoints.
+     *
+     * @var boolean
+     */
+    private $nestTransactionsWithSavepoints = false;
+
+    /**
+     * @param \Doctrine\DBAL\Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @return TransactionBuilder
+     */
+    public function createTransaction()
+    {
+        return new TransactionBuilder($this);
```

Comment:
Shouldn't the `createTransaction()` method return a transaction, not a builder?

Reply:
This is because the current implementation of the Transaction object is immutable. So the idea is to return a Builder, and only once you call `begin()`, you get a Transaction:

`createTransaction()->configureItSomehow()->begin()`.

I know that this will be confusing for some.

The other option is to make Transaction mutable *until it's begun*, but @deeky666 [didn't like this idea in my previous proposal](https://github.com/doctrine/dbal/pull/571#issuecomment-41420332).

What do you think?

Reply:
Why can't a transaction object build itself without a separate builder same as PSR-7 message objects?

Reply:
PSR 7 msges are immutable value Objects. Transaction objects cant be immutable because it has to track the transaction state (Started, running, commited, rolled back)

Reply:
Yes, you have to configure it *before* `begin()`. After it's begun, the configuration is fixed, and only the state can change (committed, rolled back).

Another solution is to create a `TransactionConfiguration` object, so that you can just do `beginTransaction($configuration)`: no more non-started Transaction objects, they could only be in one of 3 states: running, committed, rolled back.

This could actually be better when you need to start multiple transactions with the same configuration. You wouldn't have to copy/paste the transaction configuration code as you would have to with `createTransaction()->withIsolationLevel(...)->begin()`.

Reply:
As it was said below, the isolation level is the only transaction parameter SQL2003 supports. Why design a builder or configuration object which will have one value? Why not define it just as `createTransaction(int $isolationLevel)`.

Reply:
To be able to extend the Transaction object configuration in the ORM, as suggested in doctrine/doctrine2#949. In DBAL I agree that the isolation level is the only possible parameter, but in the ORM the idea was to be able to configure a default lock mode for entities loaded during this transaction only.

Reply:
In the ORM, there could be whatever ORM-specific API which would call another API of the DBAL internally. They don't have to be the same.